src/sparse_int_set.jl

```diff
@@ -0,0 +1,258 @@
+    return s
+end
+
+intersect!(s1::SparseIntSet, ss...) = intersect!(s1, intersect(ss...))
+
+#Is there a more performant way to do this?
+function intersect!(s1::SparseIntSet, ns)
+	s = intersect(s1, ns)
+	s1.packed= s.packed
+	s1.reverse = s.reverse
+	return s1
+end
+
+setdiff(s::SparseIntSet, ns) = setdiff!(copy(s), ns)
+setdiff!(s::SparseIntSet, ns) = (for n in ns; pop!(s, n, nothing); end; s)
+
+function ==(s1::SparseIntSet, s2::SparseIntSet)
+	length(s1) != length(s2) && return false
+	return all(x -> in(x, s1), s2)
+end
+
+issubset(a::SparseIntSet, b::SparseIntSet) = isequal(a, intersect(a, b))
+
+complement(a::SparseIntSet) = complement!(SparseIntSet(), a)
+function complement!(b::SparseIntSet, a::SparseIntSet)
+	empty!(b)
+	for i in eachindex(a.reverse)
+		if !isassigned(a.reverse, i)
+			resize!(b.reverse, i)
+			new_ids = (i-1)*INT_PER_PAGE+1:(i)*INT_PER_PAGE
+			append!(b.packed, new_ids)
+			b.reverse[i] = collect(new_ids)
+		else
+			for offset in 1:INT_PER_PAGE
+				if a.reverse[i][offset] == 0
+					push!(b, INT_PER_PAGE*(i-1) + offset)
+				end
+			end
+		end
+	end
+	return b
+end
+#Can this be optimized?
+complement!(a::SparseIntSet) = (b = complement(a); a.packed = b.packed; a.reverse = b.reverse; return a)
+
+<(a::SparseIntSet, b::SparseIntSet) = (a<=b) && !isequal(a,b)
+<=(a::SparseIntSet, b::SparseIntSet) = issubset(a, b)
+
+function findfirst_packed_id(i, s::SparseIntSet)
+	pageid, offset = pageid_offset(s, i)
+	if isassigned(s.reverse, pageid)
+		@inbounds id = s.reverse[pageid][offset]
+		return id
+	end
+	return 0
+end
+
+function cleanup!(s::SparseIntSet)
+	isused = x -> isassigned(s.reverse, x) && any(y -> y != 0, s.reverse[x])
+	indices = eachindex(s.reverse)
+	last_page_id = findlast(isused, indices)
+	if last_page_id === nothing
+		s.reverse = Vector{Int}[]
+		return s
+	else
+		new_pages    = Vector{Vector{Int}}(undef, last_page_id)
+		for i in indices
+			if isused(i)
+				new_pages[i] = s.reverse[i]
+			end
+		end
+		s.reverse = new_pages
+		return s
+	end
+end
+
+collect(s::SparseIntSet) = s.packed
+
+mutable struct ZippedSparseIntSetIterator{VT,IT}
+	current_id::Int
+	valid_sets::VT
+	shortest_set::SparseIntSet
+	excluded_sets::IT
+	function ZippedSparseIntSetIterator(valid_sets::SparseIntSet...;exclude::NTuple{N, SparseIntSet}=()) where{N}
+		shortest = valid_sets[findmin(map(x->length(x), valid_sets))[2]]
+		new{typeof(valid_sets), typeof(exclude)}(zero(eltype(shortest)), valid_sets, shortest, exclude)
+	end
+end
+
+zip(s::SparseIntSet...;kwargs...) = ZippedSparseIntSetIterator(s...;kwargs...)
+
+@inline length(it::ZippedSparseIntSetIterator) = length(it.shortest_set)
+
+in_excluded(id, it::ZippedSparseIntSetIterator{VT,Tuple{}}) where {VT} = false
+
+function in_excluded(id, it)
+	for e in it.excluded_sets
+		if id in e
+			return true
+		end
+	end
+	return false
+end
+
+@inline id_tids(it, state) = (id = it.shortest_set.packed[state]; return id, map(x -> findfirst_packed_id(id, x), it.valid_sets))
```

Comment:
Is this line <90 characters?
Seems really long
Might want to use a long form function